server/src/main/java/io/crate/metadata/doc/DocIndexMetadata.java

```diff
@@ -304,6 +305,10 @@ private RowGranularity granularity(ColumnIdent column) {
         return RowGranularity.DOC;
     }
 
+    public Map<String, Reference> referencesByOid() {
```

Comment:
Why is the oid a string here instead of `Long` ?

Reply:
As mentioned above:
> Also, oid key is string to avoid per-row cast of String_from_source to Long_in_map_created_once. Added javadoc to avoid confusion why oid is not Long

we do cast of long to string once on table creation and then avoid casting on parsing the source

Reply:
I don't think it is warranted to do such micro-optimizations. `DocIndexMetadata` shouldn't know about implementation details of the source parsing, but instead expose the data in a rich format.

In this case I think `DocIndexMetadata` shouldn't even expose a `referencesByOid`, but instead let the component that creates the `SourceParser` do this once.

Reply:
Currently approach is changed to use resolver. 

String is because source key could be also unknown column name. If we try to cast source key per row, not only we have some overhead which can be easily avoided but also would hit Parsing Exceptions and wold need to catch them and assign name "as is" in this case.